codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java

```diff
@@ -554,33 +584,38 @@ final void weight(short weight) {
             DefaultStream oldParent = child.parent();
-            events.add(new ParentChangedEvent(child, oldParent));
-            notifyParentChanging(child, this);
-            child.parent = this;
 
-            if (exclusive && !children.isEmpty()) {
-                // If it was requested that this child be the exclusive dependency of this node,
-                // move any previous children to the child node, becoming grand children of this node.
-                for (DefaultStream grandchild : removeAllChildren().values()) {
-                    child.takeChild(grandchild, false, events);
+            if (oldParent != this) {
+                events.add(new ParentChangedEvent(child, oldParent));
+                notifyParentChanging(child, this);
+                child.parent = this;
+                // We need the removal operation to happen first so the prioritizableForTree for the old parent to root
+                // path is updated with the correct child.prioritizableForTree() value.
+                if (oldParent != null && oldParent.children.remove(child.id()) != null) {
+                    oldParent.totalChildWeights -= child.weight();
+                    if (!child.isDescendantOf(oldParent)) {
+                        oldParent.decrementPrioritizableForTree(child.prioritizableForTree());
+                        if (oldParent.prioritizableForTree() == 0) {
+                            removeStream(oldParent);
+                        }
+                    }
                 }
-            }
 
-            // Lazily initialize the children to save object allocations.
-            initChildrenIfEmpty();
+                // Lazily initialize the children to save object allocations.
+                initChildrenIfEmpty();
+
+                final Http2Stream oldChild = children.put(child.id(), child);
+                assert oldChild == null : "A stream with the same stream ID was already in the child map.";
 
-            if (children.put(child.id(), child) == null) {
                 totalChildWeights += child.weight();
                 incrementPrioritizableForTree(child.prioritizableForTree(), oldParent);
             }
 
-            if (oldParent != null && oldParent.children.remove(child.id()) != null) {
-                oldParent.totalChildWeights -= child.weight();
-                if (!child.isDescendantOf(oldParent)) {
-                    oldParent.decrementPrioritizableForTree(child.prioritizableForTree());
-                    if (oldParent.prioritizableForTree() == 0) {
-                        removeStream(oldParent);
-                    }
+            if (exclusive && !children.isEmpty()) {
+                // If it was requested that this child be the exclusive dependency of this node,
+                // move any previous children to the child node, becoming grand children of this node.
+                for (DefaultStream grandchild : retain(child).values()) {
```

Comment:
@nmittler - Your comment was collapsed:

> > I wonder if we really need a second hash map for this. To reduce object allocation, could we just do something like:

> > DefaultStream[] childArray = children.values();
for(int index=0; index < childArray.length; ++index) {
    DefaultStream grandChild = childArray[index];
    if (child.id() != grandChild.id()) {
         child.takeChild(grandChild, false, events);
    }
}

my response:
> I think this would lead to concurrent modification scenarios.  Unless you are implying a copy operation via the `children.values();`.